pkg/querier/http.go

```diff
@@ -487,22 +487,30 @@ func WrapQuerySpanAndTimeout(call string, q *QuerierAPI) middleware.Interface {
 	return middleware.Func(func(next http.Handler) http.Handler {
 		return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
 			log, ctx := spanlogger.New(req.Context(), call)
-			userID, err := tenant.TenantID(ctx)
+			tenantIDS, err := tenant.TenantIDs(ctx)
 			if err != nil {
 				level.Error(log).Log("msg", "couldn't fetch tenantID", "err", err)
 				serverutil.WriteError(httpgrpc.Errorf(http.StatusBadRequest, err.Error()), w)
 				return
 			}
 
-			// Enforce the query timeout while querying backends
-			queryTimeout := q.limits.QueryTimeout(userID)
-			// TODO: remove this clause once we remove the deprecated query-timeout flag.
-			if q.cfg.QueryTimeout != 0 { // querier YAML configuration.
-				level.Warn(log).Log("msg", "deprecated querier:query_timeout YAML configuration identified. Please migrate to limits:query_timeout instead.", "call", "WrapQuerySpanAndTimeout")
-				queryTimeout = q.cfg.QueryTimeout
+			var longestTimeout time.Duration
+			// use the longest timeout across all tenants
+			for _, tenantID := range tenantIDS {
+				// Enforce the query timeout while querying backends
+				queryTimeout := q.limits.QueryTimeout(tenantID)
+				// TODO: remove this clause once we remove the deprecated query-timeout flag.
+				if q.cfg.QueryTimeout != 0 { // querier YAML configuration.
+					level.Warn(log).Log("msg", "deprecated querier:query_timeout YAML configuration identified. Please migrate to limits:query_timeout instead.", "call", "WrapQuerySpanAndTimeout", "org_id", tenantID)
+					queryTimeout = q.cfg.QueryTimeout
+				}
+				if longestTimeout < queryTimeout {
+					longestTimeout = queryTimeout
+
```

Comment:
small nit: this might be extra

Reply:
Sorry for testing your patience but you should use [SmallestPositiveNonZeroDurationPerTenant](https://github.com/grafana/loki/blob/main/pkg/util/validation/limits.go#L683)

Reply:
No worries and good call. I pushed https://github.com/grafana/loki/pull/7708/commits/cd034e33b3335b4df2004805a89031a10acc7145 to address this, WDYT?